lib/Service/StorageService.php

```diff
@@ -0,0 +1,231 @@
+     *
+     * @throws NotFoundException
+     * @throws \OCP\Files\InvalidPathException
+     */
+	public function createUpload(string $path, string $fileName, int $size): array
+	{
+        $currentUser = $this->userSession->getUser();
+        $userFolder = $this->rootFolder->getUserFolder(userId: $currentUser ? $currentUser->getUID() : 'Guest');
+
+        $uploadFolder = $userFolder->get($path);
+
+        $partSize = $this->config->getValueInt('openconnector', 'part-size', 1000000);
+
+        $numParts = ceil($size / $partSize);
+
+        $remainingSize = $size;
+        $parts = [];
+
+        $target = $uploadFolder->newFile($fileName);
+
+        $partsFolder = $uploadFolder->newFolder("{$fileName}_parts");
+
+
+        for ($i = 0; $i < $numParts; $i++) {
+            $partNumber = $i + 1;
+            $partUuid   = Uuid::v4();
+
+            $this->cache->set("upload_$partUuid", [
+                self::UPLOAD_TARGET_ID => $target->getId(),
+                self::UPLOAD_TARGET_PATH => $partsFolder->getPath(),
+                self::NUMBER_OF_PARTS => $numParts,
+            ]);
+
+            $parts[] = [
+                'id'    => $partUuid,
+                'size'  => $partSize < $remainingSize ? $partSize : $remainingSize,
+                'order' => $partNumber,
+            ];
+            $remainingSize -= $partSize;
+        }
+
+        return $parts;
+	}
+
+    /**
+     * Write a file to a specified path.
+     *
+     * @param string $path The path to write the file to.
+     * @param string $fileName The filename of the file to write.
+     * @param string $content The content of the file.
+     * @return File The resulting file.
+     *
+     * @throws NotFoundException
+     * @throws NotPermittedException
+     * @throws \OC\User\NoUserException
+     */
+    public function writeFile(string $path, string $fileName, string $content): File
+    {
+        $currentUser = $this->userSession->getUser();
+        $userFolder = $this->rootFolder->getUserFolder(userId: $currentUser ? $currentUser->getUID() : 'Guest');
+
+        $uploadFolder = $userFolder->get($path);
+
+        try{
+            $target = $uploadFolder->get($fileName);
+        } catch (NotFoundException $e) {
+            $target = $uploadFolder->newFile($fileName, $content);
+        }
+
+        return $target;
+    }
+
+    /**
+     * Reconcile partial files into one file if all parts of a file are present.
+     *
+     * @param Node[] $folderContents The contents of the folder containing the partial files.
+     * @param File $target The file to write the contents to.
+     *
+     * @return bool Whether or not reconciling the file has been successful.
+     */
+    private function attemptCloseUpload (array $folderContents, File $target, int $numParts): bool
+    {
+        $contentFilenames = array_map(function ($node) {
+            return $node->getName();
+        }, $folderContents);
+
+        $folder = $folderContents[0]->getParent();
```

Comment:
Can this ever be empty, what happens then?

Reply:
The moment this function is called, folderContents contains at least _n_ files where _n_ is the number of parts of the upload, so this can never be empty.